lib/irb/type_completion/type_analyzer.rb

```diff
@@ -0,0 +1,1168 @@
+# frozen_string_literal: true
+
+require 'set'
+require_relative 'types'
+require_relative 'scope'
+require 'prism'
+
+module IRB
+  module TypeCompletion
+    class TypeAnalyzer
+      class DigTarget
+        def initialize(parents, receiver, &block)
+          @dig_ids = parents.to_h { [_1.__id__, true] }
+          @target_id = receiver.__id__
+          @block = block
+        end
+
+        def dig?(node) = @dig_ids[node.__id__]
+        def target?(node) = @target_id == node.__id__
+        def resolve(type, scope)
+          @block.call type, scope
+        end
+      end
+
+      OBJECT_METHODS = {
+        to_s: Types::STRING,
+        to_str: Types::STRING,
+        to_a: Types::ARRAY,
+        to_ary: Types::ARRAY,
+        to_h: Types::HASH,
+        to_hash: Types::HASH,
+        to_i: Types::INTEGER,
+        to_int: Types::INTEGER,
+        to_f: Types::FLOAT,
+        to_c: Types::COMPLEX,
+        to_r: Types::RATIONAL
+      }
+
+      def initialize(dig_targets)
+        @dig_targets = dig_targets
+      end
+
+      def evaluate(node, scope)
+        method = "evaluate_#{node.type}"
+        if respond_to? method
+          result = send method, node, scope
```

Comment:
This looks very much like a visitor pattern. Can we use `Prism::Visitor` instead?

Reply:
Yes, it's like a visitor pattern, similar to Prism::Compiler that visit returns a value.
But analyzing also needs different setup and different argument for each child visit, needs to control visit order.

Implementing arguments to visitor will look like this
```ruby
def visit_with_args(node, args)
  @args_stack << args
  visit(node)
  @args_stack.pop
end

def visit_foobar_node(node)
  args = @args_stack.last
  v0 = calculate_something(node.child_nodes[0].name) # sometimes it needs to skip recursive visit for some node
  setup_for_2
  v1 = visit_with_arg(node.child_nodes[2], args_for_2)
  setup_for_1
  v2 = visit_with_arg(node.child_nodes[1], args_for_1)
  return calculate_something(v0, v1, v2, args)
end
```

Needs setup and arguments:
for IfNode, it needs setup, teardown and different varaible-scope-branch arguments for predicate, statements and consequent.

Needs to control visit order:
for MultiWrite node, it should visit receiver-of-write-target first, then visit value part, then visit write-targets with values to assign.

In this case, I think using visitor with stack is not as intuitive as a simple recursive visitor-like implementation.

Reply:
Hopefully this will help: https://github.com/ruby/prism/pull/1740

Reply:
Actually I'm closing that PR because @eregon had a much better suggestion, see here: https://github.com/ruby/prism/pull/1740#issuecomment-1781426941

Reply:
Additionally using Prism::Compiler should be quite a bit faster because it avoids the `respond_to?` and `send` with a dynamic method name.

> Needs to control visit order:

A visitor/Prism::Compiler can do that, you can choose the order in which you call `visit` on the children (or not call `visit` at all and handle the children in the same method). There is no need to use `visit_all`/`visit_child_nodes`.

Reply:
Thanks for the information. I like visit_with. It looks possible to implement with recreating visitor.

However, I think some of the pros of visitor is lost in this case.
I think simple dispatch with `send if respond_to?` is enough, and for this case, using visitor might introduce complexity, so I want to let this implementation as is.

Reply:
> However, I think some of the pros of visitor is lost in this case.

> using visitor might introduce complexity

Could you explain with examples to help me get the context?

Reply:
Benefit is that the implementation is simple.

These pros of using Visitor are lost in this case.
- Only need to implement few method focusing on: in this case, we need to implement every single one.
- Does not need to know how to traverse: in this case, we need to specify traverse order for every node.
- Fast: Recreating visitor will make it slightly slow (and make code complex)
- Double dispatch mechanism: Dispatch is easy in ruby compared to other language.

If the pros is relatively small, using visitor will just make the code complicated.

This is the current implementation
```ruby
def evaluate(node, scope); dispatch using send; end

def evaluate_foo_node(node, scope)
  scope2 = scope.new_child_scope
  a = evaluate(node.a, scope2)
  b = evaluate(node.b, scope2)
  scope.update(scope2)
  type_of_foo_calculation(a,b)
end
```

Using visitor will look like this
```ruby
def visit_foo_node(node)
  scope2 = @scope.new_child_scope
  a = self.class.new(@bar, scope2).visit(node.a) # boilerplate to visit with args
  b = self.class.new(@bar, scope2).visit(node.b)
  @scope.update(scope2)
  type_of_foo_calculation(a,b)
end
```

The boilerplate can be cleared by this
```ruby
def evaluate(node, scope)
  self.class.new(@bar, scope).visit(node)
end

def visit_foo_node(node)
  scope2 = @scope.new_child_scope
  a = evaluate(node.a, scope2)
  b = evaluate(node.b, scope2)
  @scope.update(scope2)
  type_of_foo_calculation(a,b)
end
```

But as you see, it's almost same to current implementation. It's essentially same but with some complexity overhead.
This is just using visitor as method-dispatching that can be also obtained by `send`. Using Visitor class but not using its traverse mechanism is confusing.
`self` might be changed while visiting nodes. It means that setting instance variable has no effect to the root visitor instance and the fact should be recognized while maintaining.
`node` and `@scope` are calculation parameter of `visit_foo_node(node)` but `@scope` is not in the method args. In the current implementation, `evaluate_foo_node(node, scope)` explicitly represents it.